// checkly/dist/commands/env/add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const api = require("../../rest/api");
const core_1 = require("@oclif/core");
const authCommand_1 = require("../authCommand");
class EnvAdd extends authCommand_1.AuthCommand {
    async run() {
        var _a;
        const { flags, args } = await this.parse(EnvAdd);
        const { locked } = flags;
        const envVariableName = args.key;
        let envValue = '';
        // check if env variable exists
        if (args.value) {
            envValue = args.value;
        }
        else {
            envValue = await core_1.ux.prompt(`What is the value of ${envVariableName}?`, { type: 'mask' });
        }
        try {
            await api.environmentVariables.add(envVariableName, envValue, locked);
            this.log(`Environment variable ${envVariableName} added.`);
        }
        catch (err) {
            if (((_a = err === null || err === void 0 ? void 0 : err.response) === null || _a === void 0 ? void 0 : _a.status) === 409) {
                throw new Error(`Environment variable ${envVariableName} already exists.`);
            }
            throw err;
        }
    }
}
exports.default = EnvAdd;
EnvAdd.hidden = false;
EnvAdd.description = 'Add environment variable via "checkly env add <key> <value>".';
EnvAdd.flags = {
    locked: core_1.Flags.boolean({
        char: 'l',
        description: 'Indicate that the environment variable will be locked.',
        default: false,
    }),
};
EnvAdd.args = {
    key: core_1.Args.string({
        name: 'arguments',
        required: true,
        description: 'Environment variable key.',
    }),
    value: core_1.Args.string({
        name: 'arguments',
        required: false,
        description: 'Environment variable value.',
    }),
};
